src/client.ts

```diff
@@ -1,149 +1,224 @@
+      throw new IntegrationProviderAuthenticationError({
+        cause: new Error('Provider authentication failed'),
+        endpoint: groupsApiRoute,
+        status: response.status,
+        statusText: response.statusText,
       });
     }
   }
 
   /**
-   * Iterates each user resource in the provider.
+   * Iterates each group in the provider.
    *
    * @param iteratee receives each resource to produce entities/relationships
    */
-  public async iterateUsers(
-    iteratee: ResourceIteratee<AcmeUser>,
+  public async iterateGroups(
+    iteratee: ResourceIteratee<MalwareBytesGroup>,
   ): Promise<void> {
-    // TODO paginate an endpoint, invoke the iteratee with each record in the
-    // page
-    //
-    // The provider API will hopefully support pagination. Functions like this
-    // should maintain pagination state, and for each page, for each record in
-    // the page, invoke the `ResourceIteratee`. This will encourage a pattern
-    // where each resource is processed and dropped from memory.
-
-    const users: AcmeUser[] = [
-      {
-        id: 'acme-user-1',
-        name: 'User One',
-      },
-      {
-        id: 'acme-user-2',
-        name: 'User Two',
-      },
-    ];
+    const response = await this.request(this.withBaseUri('nebula/v1/groups'));
+    const groupsResponse = await response.json();
 
-    for (const user of users) {
-      await iteratee(user);
+    for (const group of groupsResponse.groups) {
+      await iteratee(group);
     }
   }
 
   /**
-   * Iterates each group resource in the provider.
+   * Iterates each endpoint found within the requested timespan for a given group.
    *
+   * @param groupName group's name
    * @param iteratee receives each resource to produce entities/relationships
    */
-  public async iterateGroups(
-    iteratee: ResourceIteratee<AcmeGroup>,
+  public async iterateEndpoints(
+    groupName: string,
+    iteratee: ResourceIteratee<MalwareBytesEndpointWithAgent>,
   ): Promise<void> {
-    // TODO paginate an endpoint, invoke the iteratee with each record in the
-    // page
-    //
-    // The provider API will hopefully support pagination. Functions like this
-    // should maintain pagination state, and for each page, for each record in
-    // the page, invoke the `ResourceIteratee`. This will encourage a pattern
-    // where each resource is processed and dropped from memory.
-
-    const groups: AcmeGroup[] = [
-      {
-        id: 'acme-group-1',
-        name: 'Group One',
-        users: [
-          {
-            id: 'acme-user-1',
-          },
-        ],
-      },
-    ];
+    const endpoints = await this.makePaginatedRequest<
+      'endpoints',
+      MalwareBytesEndpointWithAgent
+    >(this.withBaseUri('nebula/v1/endpoints'), 'POST', 'endpoints', {
+      'group_name.keyword': groupName,
+      last_day_seen_after: this.config.minScannedSinceISODate,
+      last_day_seen_before: this.config.maxScannedSinceISODate,
+    });
```

Comment:
I think it would be great if these request body properties were generated on the fly from `context.instance.config` and `context.executionHistory`. I would instantiate the `new ApiClient()` with `config` and `executionHistory` properties, and make them instance variables.

```suggestion
    const lastDaySeenAfter = getLastDaySeenAfter({ minScannedSinceDays: this.config. minScannedSinceDays, lastSuccessful: this.executionHistory.lastSuccessful });
    const lastDaySeenBefore = getLastDaySeenBefore({ startedOn: this.executionHistory.current.startedOn, });

    const endpoints = await this.makePaginatedRequest<
      'endpoints',
      MalwareBytesEndpointWithAgent
    >(this.withBaseUri('nebula/v1/endpoints'), 'POST', 'endpoints', {
      'group_name.keyword': groupName,
      last_day_seen_after: lastDaySeenAfter,
      last_day_seen_before: lastDaySeenBefore,
    });
```

Reply:
First of all, thanks a lot for providing the suggested change - it helped a lot!

In the changed code, you'll see the following:
```
const lastDaySeenAfter = getLastDaySeenAfter({
      minSinceDays: this.config.minScannedSinceDays
        ? parseInt(this.config.minScannedSinceDays, 10)
        : DEFAULT_SCANNED_SINCE_DAYS,
      lastSuccessful: this.executionHistory.lastSuccessful,
    });
```

Reasons behind it are:

1. If the value is coming from ENV, it's a string - so we convert it to a number.
2. We're using the `DEFAULT_SCANNED_SINCE_DAYS` value here (as opposed inside the `getLastDaySeenAfter()`) because we wanted to keep `getLastDaySeenAfter()` as generic as possible (so that it can be used for both endpoints and findings). If we wanted to do `|| DEFAULT_VALUE` inside of it, we'd need to also pass something to denote if the current call is about findings or endpoints to know which default value to use.
3. For similar reasons, we're calling the property `minScannedSinceDays` just `minSinceDays`.

Let me know if this is fine or if you've envisioned something slightly different.

Reply:
I like those changes! Thanks for also providing your reasoning, very clear to follow 👍